[dagster-embedded-elt][dlt] Documentation for dlt integration

cmpadden opened the pull request:
Adds a section to _Embedded ELT_ on `dlt`, including an example on how to pull a `dlt` verified source, and add it as a Dagster `dlt_assets`.

cmpadden commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20763?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#20763** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20763?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#20759** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20759?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @cmpadden and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

PedramNavid commented on docs/content/integrations/embedded-elt.mdx:
```diff
@@ -7,251 +7,11 @@ description: Lightweight ELT framework for building ELT pipelines with Dagster,
+1. <a href="https://slingdata.io">Sling</a> to provide a simple way to sync data
+   between databases and file systems.
+2. <a href="https://dlthub.com">Data Load Tool (dlt)</a> to easily load data from
+   external systems and APIs.
```
Should link to the two sub-pages here as well.

cmpadden replied:
Added links with short descriptions here: https://github.com/dagster-io/dagster/pull/20763/commits/24fd70ad1548d55893be2581be8fefc84660f24b

PedramNavid commented on docs/content/integrations/embedded-elt/dlt.mdx:
```diff
@@ -0,0 +1,127 @@
+
+To get started with `dagster-embedded-elt` and _dlt_, first familiarize yourself with the quick [introduction to dlt](https://dlthub.com/docs/intro). When using the _dlt_ integration, you can either write your own sources, or use one of the many pre-made [verified sources](https://dlthub.com/docs/dlt-ecosystem/verified-sources/) in the _dlt_ ecosystem.
+
+## 1. Configuring your Dagster project to support `dlt`
```
I think this jumps too quickly into 'tutorial' mode. For people who are not wanting to do a tutorial, we should do a better job explaining the high level concepts they should be familiar with.

cmpadden replied:
Thank you - I added a background section, but it may need more. We'll see!

PedramNavid commented on docs/content/integrations/embedded-elt/dlt.mdx:
````diff
@@ -0,0 +1,127 @@
+
+This has pulled down the code required to easily collect data from the GitHub API. It has also created a `requirements.txt` and a `.dlt/` configuration. In this case, we can remove these files, as we will be handling configuration with environment variables, and the requirements are to be defined in Dagster's dependencies specification.
+
+```bash
````
wonder if it's better to just do the `rm` command rather than show the files?

cmpadden replied:
I think it may be beneficial to keep it, as you can reference the requirements.txt and `.dlt/secrets.yaml` to understand what needs to be added elsewhere in the project.

PedramNavid commented on docs/content/integrations/embedded-elt/dlt.mdx:
```diff
@@ -0,0 +1,127 @@
+
+This integration manages connections and secrets using environment variables as `dlt`. The `dlt` library is able to "magically" infer required environment variables used by its sources and resources. See [Secrets and Configs](https://dlthub.com/docs/general-usage/credentials/configuration) for more information.
+
+In this example, the `github_reactions` source requires a GitHub access token, and the _snowflake_ destination requires database connection details. This results in the required environment variables:
```
`This results in the following required environment variables:`

PedramNavid commented on docs/content/integrations/embedded-elt/dlt.mdx:
````diff
@@ -0,0 +1,127 @@
+DESTINATION__SNOWFLAKE__CREDENTIALS__ROLE=""
+```
+
+Ensure that these are defined in your `.env` file, and in the Dagster Cloud environment.
````
```suggestion
Ensure that these are defined in your environment
```

PedramNavid commented on docs/content/integrations/embedded-elt/dlt.mdx:
```diff
@@ -0,0 +1,127 @@
+
+## 4. Define a `DagsterDltResource`
+
+Create a `DagsterDltResource` resources:
```
```suggestion
Create a `DagsterDltResource` resource:
```

PedramNavid commented on docs/content/integrations/embedded-elt/dlt.mdx:
```diff
@@ -0,0 +1,127 @@
+Ensure that these are defined in your `.env` file, and in the Dagster Cloud environment.
+
+## 4. Define a `DagsterDltResource`
+
```
Why am I doing this?

PedramNavid commented on docs/content/integrations/embedded-elt/dlt.mdx:
```diff
@@ -0,0 +1,127 @@
+
+## 5. Create a `dlt_assets` definition for GitHub
+
+Using the `dlt_assets` decorator, we can specify the `dlt` GitHub source as the `dlt_source` parameter, and our pipeline specifying that we would like the data loaded to Snowflake in the `dlt_pipeline`. In the body of the decorated function we call the `dlt.run`
```
```suggestion
The `dlt_assets` decorator takes a `dlt_source` and `dlt_pipeline` parameter. In this example, we will use the `github_reactions` source and create a `dlt_pipeline` to ingest from Github to Snowflake. 

We then yield from `dlt.run` in the body of the function which will execute the pipeline when this asset is materialized.
```

cmpadden commented:
> A few comments, I am sure @erinkcochran87 will have more :)

Thanks for the review, @PedramNavid . I believe I've addressed your comments. Ready for the next iteration from Erin, heh.

erinkcochran87 commented on docs/content/integrations/embedded-elt/dlt.mdx:
````diff
@@ -30,57 +40,88 @@ def example(api_key: str = dlt.secrets.value):
-pip3 install -r requirements.txt
-* Read https://dlthub.com/docs/walkthroughs/create-a-pipeline for more information
-```
+   ```bash
+   pip install dlt
+   ```
````
I added a prereqs section to this, which is standard for our guides. I feel like it would be better to add it in with installing the other libraries - what do you think?

cmpadden replied:
Actually, once a user installs `dagster-embedded-elt` `dlt` will be installed automatically. I will remove this item.

erinkcochran87 replied:
Oh good call! I keep forgetting we do this with our integration libraries.

erinkcochran87 commented on docs/content/integrations/embedded-elt/dlt.mdx:
```diff
@@ -100,11 +141,18 @@ $ tree -a
+- The `github_reactions` source requires a GitHub access token
+- The Snowflake destination requires database connection details
+
+This results in the following required environment variables:
```
Where should users be defining these? In an `.env`? Elsewhere? We should be specific here if we can.

erinkcochran87 commented on docs/content/integrations/embedded-elt/dlt.mdx:
```diff
@@ -116,21 +164,25 @@ DESTINATION__SNOWFLAKE__CREDENTIALS__WAREHOUSE=""
+## Step 4: Define a DagsterDltResource
 
-The <PyObject module="dagster_embedded_elt.dlt" object="DagsterDltResource" /> provides a wrapper of a _dlt_ pipeline runner. This can be shared across _dlt_ pipelines, and is a required parameter to our _dlt_ asset definition.
+The <PyObject module="dagster_embedded_elt.dlt" object="DagsterDltResource" /> provides a wrapper of a dlt pipeline runner. This can be shared across dlt pipelines and is a required parameter to our dlt asset definition.
```
What file should users be doing this in?

cmpadden replied:
This one I wasn't sure of, because it would be in a hypothetical `resource.py` file, or wherever the user typically defines their resources. In that sense, this "tutorial" isn't so much a tutorial as we don't prescribe the file structure. This could be changed, though.

erinkcochran87 commented on docs/content/integrations/embedded-elt/dlt.mdx:
```diff
@@ -116,21 +164,25 @@ DESTINATION__SNOWFLAKE__CREDENTIALS__WAREHOUSE=""
+## Step 5: Create a dlt_assets definition for GitHub
 
-The `dlt_assets` decorator takes a `dlt_source` and `dlt_pipeline` parameter. In this example, we will use the `github_reactions` source and create a `dlt_pipeline` to ingest from Github to Snowflake.
+The `dlt_assets` decorator takes a `dlt_source` and `dlt_pipeline` parameter. In this example, we used the `github_reactions` source and created a `dlt_pipeline` to ingest data from Github to Snowflake:
```
What file should users be doing this in?

Also, we should show the imports being used as well. Code snippets should be as copy-and-pasteable as possible.

erinkcochran87 reviewed (commented):
Back to you. A few things:

- You can use Markdown links (`[text](url)`) instead of `<a href`
- Our style guide doesn't use any formatting in headings, ex: **Use the `SlingResource`** should just be **Use the SlingResource**
- We also don't need to style the names of integrations - ex: `_dlt_` should just be `dlt`

cmpadden commented:
> Back to you. A few things:
> 
> * You can use Markdown links (`[text](url)`) instead of `<a href`
> * Our style guide doesn't use any formatting in headings, ex: **Use the `SlingResource`** should just be **Use the SlingResource**
> * We also don't need to style the names of integrations - ex: `_dlt_` should just be `dlt`
> 
> I changed these already, just noting them here for your next docs adventure!

Thanks a bunch, @erinkcochran87 - I believe I've addressed your feedback. I've also included the integration in our Sphinx `rst` file, I missed that at first. Let me know if there's anything else.

cmpadden commented:
@erinkcochran87 whew, I got stuck in merge conflict troubles there for a bit, I think the PR is back in a good state.

erinkcochran87 reviewed (approved):
@cmpadden This looks good! I did a last round of editing so we can get this out the door, so you'll need to pull everything if you have any other changes to make. 

I think there's a BK error to address, but otherwise this is gtg 🚀
